File: cpp/binary-search-tree/binary_search_tree.cpp
#include "binary_search_tree.h"

namespace binary_search_tree {

    template<class T>
    binary_tree<T>::binary_tree() = default;

    template<class T>
    binary_tree<T>::binary_tree(T value) : value(value), size(1) {}

    template<class T>
    binary_tree<T>::~binary_tree() {
        delete this->left_c;
        delete this->right_c;
    }

    template<class T>
    void binary_tree<T>::insert(T data) {
        if (data < this->value)
            if (this->left_c) this->left_c->insert(data);
            else this->left_c = new binary_tree(data);
        else
            if (this->right_c) this->right_c->insert(data);
            else this->right_c = new binary_tree(data);
        this->size += 1;
    }

    template<class T>
    std::unique_ptr<binary_tree<T>> binary_tree<T>::left() const {
        return std::unique_ptr(this->left_c);
    }

    template<class T>
    std::unique_ptr<binary_tree<T>> binary_tree<T>::right() const {
        return std::unique_ptr(this->right_c);
    }

    template<class T>
    T binary_tree<T>::data() const {
        return this->value;
    }

    template<class T>
    T *binary_tree<T>::begin() {
        size_t idx = 0;
        auto arr = new T[this->size];
        this->fill_array(arr, idx);
        return arr;
    }

    template<class T>
    T *binary_tree<T>::end() {
        return &this->begin()[this->size];
    }

    template<class T>
    void binary_tree<T>::fill_array(T *arr, size_t &idx) const {
        if (this->left_c) this->left_c->fill_array(arr, idx);
        arr[idx++] = this->value;
        if (this->right_c) this->right_c->fill_array(arr, idx);
    }
}  // namespace binary_search_tree

File: java/binary-search-tree/src/main/java/BinarySearchTree.java
import java.util.*;

class BinarySearchTree<T extends Comparable<T>> {
    private Node<T> root = null;

    void insert(T value) {
        if (root == null) {
            root = new Node<>(value);
        } else {
            root.insert(value);
        }
    }

    List<T> getAsSortedList() {
        List<T> list = new ArrayList<>();
        if (root != null)
            root.appendToSortedList(list);
        return list;
    }

    List<T> getAsLevelOrderList() {
        List<T> list = new ArrayList<>();
        Queue<Node<T>> queue = new LinkedList<>();
        if (root != null) {
            queue.add(root);
            while (queue.size() > 0) {
                var node = queue.remove();
                if (node.left != null) queue.add(node.left);
                if (node.right != null) queue.add(node.right);
                list.add(node.data);
            }
        }
        return list;
    }

    Node<T> getRoot() {
        return root;
    }

    static class Node<T extends Comparable<T>> {
        private final T data;
        private Node<T> left = null;
        private Node<T> right = null;

        Node(T data) {
            this.data = data;
        }

        Node<T> getLeft() {
            return left;
        }

        Node<T> getRight() {
            return right;
        }

        T getData() {
            return data;
        }

        void insert(T data) {
            if (data.compareTo(this.data) <= 0) {
                if (left == null)
                    left = new Node<>(data);
                else
                    left.insert(data);
            } else {
                if (right == null) {
                    right = new Node<>(data);
                } else {
                    right.insert(data);
                }
            }
        }

        void appendToSortedList(List<T> list) {
            if (left != null) left.appendToSortedList(list);
            list.add(data);
            if (right != null) right.appendToSortedList(list);
        }
    }
}
